### DISCUSSION TITLE
CORS issues with local vercel dev PATCH POST DELETE endpoints

### DISCUSSION THREAD

**Discussion by mbenfowler** [vercel/community#4794] [Help] [answered]
### Summary

I'm trying to do a simple serverless function/knex/postgres backend and am using vercel dev wanting to test locally before merging into main. But only on local, I'm running into CORS issues on just the PATCH POST and DELETE endpoints.  I am able to locally hit these endpoints via the browser url and postman and all GET endpoints work as expected from within the front end app locally and on production. 

### Steps to Reproduce

Here is my vanilla vercel.json:
```
{
  "headers": [
    {
      "source": "/api/(.*)",
      "headers": [
        { "key": "Access-Control-Allow-Credentials", "value": "true" },
        { "key": "Access-Control-Allow-Origin", "value": "*" },
        { "key": "Access-Control-Allow-Methods", "value": "GET,OPTIONS,PATCH,DELETE,POST,PUT" },
        { "key": "Access-Control-Allow-Headers", "value": "X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version" }
      ]
    }
  ]
}
```

And here is one of the endpoints running into the issue:
```
deleteSaved.js
const { configDotenv } = require("dotenv")

configDotenv()

const environment = process.env.NODE_ENV || 'development'
const configuration = require('../knexfile')[environment]
const db = require('knex')(configuration)

module.exports = async (req, res) => {
  try {
    const speciesCode = req.query.speciesCode
    const birdFromDB = await db('birds').where('speciesCode', speciesCode).first();
    if (birdFromDB) {
      await db('saved_birds').where('bird_id', birdFromDB.id).del();
      res.status(200).json(birdFromDB);
    } else {
      res.status(404).json({ error: 'Bird not found' });
    }
  } catch (error) {
    console.error("Error in deleteSaved:", error);
    res.status(500).json({ error });
  }
};
```

Here is the error in the browser console:
```
Access to fetch at 'http://localhost:3001/api/deleteSaved?speciesCode=snogoo' from origin 'http://localhost:3000' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: It does not have HTTP ok status.
apiCalls.js:125     DELETE http://localhost:3001/api/deleteSaved?speciesCode=snogoo net::ERR_FAILED
deleteSavedBird @ apiCalls.js:125
handleClick @ Card.js:22
```

And the headers from the network request:
```
Request URL:
http://localhost:3001/api/deleteSaved?speciesCode=snogoo
Request Method:
OPTIONS
Status Code:
500 Internal Server Error
Remote Address:
[::1]:3001
Referrer Policy:
strict-origin-when-cross-origin
Access-Control-Allow-Credentials:
true
Access-Control-Allow-Headers:
X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version
Access-Control-Allow-Methods:
GET,OPTIONS,PATCH,DELETE,POST,PUT
Access-Control-Allow-Origin:
*
Cache-Control:
public, max-age=0, must-revalidate
Connection:
close
Content-Length:
436
Date:
Thu, 09 Nov 2023 18:05:07 GMT
Server:
Vercel
X-Vercel-Cache:
MISS
X-Vercel-Id:
dev1::dev1::4i3to-1699553107461-3dbddd49b442
Accept:
*/*
Accept-Encoding:
gzip, deflate, br
Accept-Language:
en-US,en;q=0.9
Access-Control-Request-Headers:
content-type
Access-Control-Request-Method:
DELETE
Connection:
keep-alive
Host:
localhost:3001
Origin:
http://localhost:3000
Referer:
http://localhost:3000/
Sec-Fetch-Dest:
empty
Sec-Fetch-Mode:
cors
Sec-Fetch-Site:
same-site
User-Agent:
Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/117.0.0.0 Safari/537.36
```

**Comment by mbenfowler** [comment]
ChatGPT says I may need to create an express alternate for local testing, but I'd prefer not to do that if I don't have to:

To run a serverless function locally without using vercel dev, you can use the @vercel/node builder directly. Here's how you can do it:

1) Install Required Packages:

Make sure you have the necessary packages installed. You can install them using npm:
```
npm install -g vercel
npm install express
```

2) Create a Server File:
Create a file, let's call it `server.js`, and define your serverless functions using Express:
```
const express = require('express');
const server = express();

server.get('/api/hello', (req, res) => {
  res.json({ message: 'Hello, World!' });
});

module.exports = server;
```

3) Deploy Locally:
Deploy your serverless functions locally using the @vercel/node builder:
`vercel dev server.js`
Replace server.js with the name of your file containing the serverless functions.

4) Test Locally:
Your serverless functions should be accessible at `http://localhost:3000/api/hello` or a similar URL based on your route configuration.

This approach allows you to run serverless functions locally without using the vercel dev command. It uses the @vercel/node builder directly and provides a simple way to test your serverless functions locally.

**Comment by amyegan** [comment]
Hi @mbenfowler! Have you tried adding the headers in your Express code instead of vercel.json? 

This post has some good suggestions: https://stackoverflow.com/questions/34644622/how-to-manage-cors-policy-properly-in-express

**Reply by mbenfowler** [reply]
@amyegan thanks so much for the quick response!!

I had seen this post, but I was under the impression that the settings in `vercel.json` were supposed to handle everything that the cors middleware was doing for us?  I was also using express endpoints in a single file previously and was having worse CORS issues getting things to work locally and live, and I've been having significantly more success after switching to the serverless functions where each endpoint is in it's own file in the `/api` path.

**Comment by mbenfowler** [comment, marked as answer]
Okay I figured out a workaround:
1)  Removing the headers from the front end request fixed a couple functions.
2) I had to remove the `method` from a PATCH and DELETE for them to work.

Still interesting that it was working just fine on prod.

**Reply by huynhr** [reply]
I don't know how or why but removing the headers like you said seemed to work for me as well. The funny thing is, the browser still throws a cors preflight error 🤷🏼‍♂️ .

**Reply by TadasMil** [reply]
Removing `headers` from the client is not a solution. I found this to be working [DOCS](https://vercel.com/guides/how-to-enable-cors) :
```
export default async function handler(req: VercelRequest, res: VercelResponse) {
    if (req.method === 'OPTIONS') {
      res.setHeader('Access-Control-Allow-Origin', '*');
      res.setHeader('Access-Control-Allow-Methods', 'GET, POST, PUT, DELETE, PATCH, OPTIONS');
      res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization, X-Requested-With, Accept, X-CSRF-Token, X-Api-Version');
      return res.status(200).send('');
    }

    return res.json({ message: 'Hello World' });
  }
```
